File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/vehicle/GroundFrictionForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.vehicle;

import com.github.aadvorak.artilleryonline.battle.calculations.*;
import com.github.aadvorak.artilleryonline.battle.calculator.utils.ForceCalculatorUtils;
import com.github.aadvorak.artilleryonline.battle.common.Constants;
import com.github.aadvorak.artilleryonline.battle.common.Force;
import com.github.aadvorak.artilleryonline.battle.common.Vector;
import com.github.aadvorak.artilleryonline.battle.config.VehicleConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;
import com.github.aadvorak.artilleryonline.battle.precalc.VehiclePreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.VehicleSpecs;
import com.github.aadvorak.artilleryonline.battle.state.VehicleState;

import java.util.ArrayList;
import java.util.List;

public class GroundFrictionForceCalculator implements ForceCalculator<
        VehicleSpecs,
        VehiclePreCalc,
        VehicleConfig,
        VehicleState,
        VehicleModel,
        VehicleCalculations> {

    private static final String FORCE_DESCRIPTION = "Ground Friction";

    @Override
    public List<BodyForce> calculate(VehicleCalculations calculations, BattleModel battleModel) {
        var forces = new ArrayList<BodyForce>();
        if (calculations.getModel().getState().getVelocity().magnitude() < Constants.ZERO_THRESHOLD) {
            return forces;
        }
        var groundFrictionCoefficient = battleModel.getRoom().getSpecs().getGroundFrictionCoefficient();
        var gravityAcceleration = battleModel.getRoom().getSpecs().getGravityAcceleration();
        var contactsNumber = calculations.getGroundContacts().size();
        addWheelFriction(forces, calculations.getRightWheel(), groundFrictionCoefficient, gravityAcceleration, contactsNumber);
        addWheelFriction(forces, calculations.getLeftWheel(), groundFrictionCoefficient, gravityAcceleration, contactsNumber);
        addTurretFriction(forces, calculations, groundFrictionCoefficient, gravityAcceleration, contactsNumber);
        return forces;
    }

    private void addWheelFriction(List<BodyForce> forces, WheelCalculations wheelCalculations,
                                  double groundFrictionCoefficient, double gravityAcceleration,
                                  int contactsNumber) {
        var contact = wheelCalculations.getGroundContact();
        if (contact == null) {
            return;
        }
        var coefficient = groundFrictionCoefficient * gravityAcceleration * wheelCalculations.getMass()
                * Math.cos(contact.angle()) / contactsNumber / 10;
        var velocity = wheelCalculations.getVelocity().projectionOnto(
                Vector.tangential(contact.angle()));
        var force = new Force()
                .setX( - velocity.getX() * coefficient)
                .setY( - velocity.getY() * coefficient);
        forces.add(BodyForce.of(force, wheelCalculations.getPosition(), wheelCalculations.getModel().getState().getPosition().getCenter(),
                FORCE_DESCRIPTION + " Wheel"));
    }

    private void addTurretFriction(List<BodyForce> forces, VehicleCalculations calculations,
                                   double groundFrictionCoefficient, double gravityAcceleration,
                                   int contactsNumber) {
        if (calculations.getTurretGroundContacts() == null) {
            return;
        }
        var forceMultiplier = groundFrictionCoefficient * gravityAcceleration * calculations.getMass() / contactsNumber;
        calculations.getTurretGroundContacts().forEach(contact -> ForceCalculatorUtils.addFriction(forces,
                calculations, contact, forceMultiplier, FORCE_DESCRIPTION + " Hull"));
    }
}

File: frontend/playground/battle/calculator/vehicle/ground-friction-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import {type VehicleCalculations, type WheelCalculations} from "~/playground/data/calculations";
import type {BattleModel, VehicleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {BodyUtils} from "~/playground/utils/body-utils";

export class GroundFrictionForceCalculator implements ForceCalculator<VehicleCalculations> {
  private static readonly FORCE_DESCRIPTION = 'Ground Friction'

  calculate(calculations: VehicleCalculations, battleModel: BattleModel): BodyForce[] {
    const groundFrictionCoefficient = battleModel.room.specs.groundFrictionCoefficient
    const gravityAcceleration = battleModel.room.specs.gravityAcceleration
    const contactsNumber = BodyUtils.getAllGroundContacts(calculations).length
    const forces: BodyForce[] = []

    this.addWheelFriction(forces, calculations.rightWheel, calculations.model, groundFrictionCoefficient, gravityAcceleration, contactsNumber)
    this.addWheelFriction(forces, calculations.leftWheel, calculations.model, groundFrictionCoefficient, gravityAcceleration, contactsNumber)
    this.addHullFriction(forces, calculations, groundFrictionCoefficient, gravityAcceleration, contactsNumber)

    return forces
  }

  private addWheelFriction(
      forces: BodyForce[],
      wheelCalculations: WheelCalculations,
      vehicleModel: VehicleModel,
      groundFrictionCoefficient: number,
      gravityAcceleration: number,
      contactsNumber: number
  ): void {
    const contact = wheelCalculations.groundContact
    if (!contact) {
      return
    }
    const coefficient = groundFrictionCoefficient * gravityAcceleration
        * vehicleModel.preCalc.mass * Math.cos(contact.angle) / contactsNumber / 10
    const velocity = VectorUtils.projectionOfOnto(
        wheelCalculations.velocity!,
        VectorUtils.tangential(contact.angle)
    )

    const force = {
      x: -velocity.x * coefficient,
      y: -velocity.y * coefficient
    }

    forces.push(BodyForce.of(
        force,
        wheelCalculations.position!,
        vehicleModel.state.position,
        GroundFrictionForceCalculator.FORCE_DESCRIPTION + ' Wheel'
    ))
  }

  private addHullFriction(
      forces: BodyForce[],
      calculations: VehicleCalculations,
      groundFrictionCoefficient: number,
      gravityAcceleration: number,
      contactsNumber: number
  ): void {
    const groundContacts = calculations.groundContacts
    if (!groundContacts) {
      return
    }

    const forceMultiplier = groundFrictionCoefficient * gravityAcceleration
        * calculations.model.preCalc.mass / contactsNumber

    groundContacts.forEach(contact => {
      const tangential = VectorUtils.tangential(contact.angle)
      const movingVelocity = VectorUtils.projectionOfOnto(calculations.model.state.velocity, tangential)
      const rotatingVelocity = VectorUtils.projectionOfOnto(
          BodyUtils.getRotatingVelocityAt(calculations.model.state, contact.position), tangential
      )
      const movingVelocityMagnitude = VectorUtils.getMagnitude(movingVelocity)
      const rotatingVelocityMagnitude = VectorUtils.getMagnitude(rotatingVelocity)
      const contactForceMultiplier = forceMultiplier * Math.cos(contact.angle)

      if (movingVelocityMagnitude > rotatingVelocityMagnitude || rotatingVelocityMagnitude < 0.05) {
        const velocity = BodyUtils.getVelocityAt(calculations.model.state, contact.position)
        const movingForce = {
          x: -velocity.x * contactForceMultiplier,
          y: -velocity.y * contactForceMultiplier
        }
        forces.push(BodyForce.of(
            movingForce,
            contact.position,
            calculations.model.state.position,
            GroundFrictionForceCalculator.FORCE_DESCRIPTION + ' Hull'
        ))
      } else {
        const rotatingForce = {
          x: -rotatingVelocity.x * contactForceMultiplier,
          y: -rotatingVelocity.y * contactForceMultiplier
        }
        forces.push(BodyForce.atCOM(rotatingForce, GroundFrictionForceCalculator.FORCE_DESCRIPTION + ' Hull'))
      }
    })
  }
}
